File: 044_MergeKSortedLists.cpp
// Topic: Merge K Sorted Lists
// Description: Merge multiple sorted linked lists into one sorted list efficiently.
// Approach: Use a min-heap (priority_queue) of current nodes, O(N log k) where N total nodes.

#include <bits/stdc++.h>
using namespace std;

struct ListNode
{
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

struct Cmp
{
    bool operator()(ListNode* a, ListNode* b) const { return a->val > b->val; }
};

ListNode* mergeKLists(vector<ListNode*>& lists)
{
    priority_queue<ListNode*, vector<ListNode*>, Cmp> pq;
    for (auto l : lists)
        if (l)
            pq.push(l);
    ListNode dummy(0), *tail = &dummy;
    while (!pq.empty())
    {
        ListNode* node = pq.top();
        pq.pop();
        tail->next = node;
        tail = tail->next;
        if (node->next)
            pq.push(node->next);
    }
    return dummy.next;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    // Test 1
    {
        vector<ListNode*> lists;
        ListNode* a = new ListNode(1);
        a->next = new ListNode(4);
        a->next->next = new ListNode(5);
        ListNode* b = new ListNode(1);
        b->next = new ListNode(3);
        b->next->next = new ListNode(4);
        ListNode* c = new ListNode(2);
        c->next = new ListNode(6);
        lists.push_back(a);
        lists.push_back(b);
        lists.push_back(c);
        ListNode* res = mergeKLists(lists);
        for (ListNode* p = res; p; p = p->next)
        {
            cout << p->val;
            if (p->next)
                cout << " ";
        }
        cout << "\n";
    }
}

/*
Time Complexity: O(N log k)
Space Complexity: O(k)
*/

File: 044_MergeKSortedLists.py
# Topic: Merge K Sorted Lists
# Description: Merge multiple sorted linked lists into one sorted list efficiently.
# Approach: Use a min-heap (heapq) of current nodes, O(N log k) where N total nodes.

from typing import List, Optional
import heapq


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def __lt__(self, other):
        return self.val < other.val


class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        heap = []
        for l in lists:
            if l:
                heapq.heappush(heap, l)
        dummy = ListNode(0)
        tail = dummy
        while heap:
            node = heapq.heappop(heap)
            tail.next = node
            tail = tail.next
            if node.next:
                heapq.heappush(heap, node.next)
        return dummy.next


def print_list(head):
    vals = []
    while head:
        vals.append(str(head.val))
        head = head.next
    print(' '.join(vals))


if __name__ == "__main__":
    # Test 1
    a = ListNode(1, ListNode(4, ListNode(5)))
    b = ListNode(1, ListNode(3, ListNode(4)))
    c = ListNode(2, ListNode(6))
    result = Solution().mergeKLists([a, b, c])
    print_list(result)

"""
Time Complexity: O(N log k)
Space Complexity: O(k)
"""
